### DISCUSSION TITLE
How can I specify that a value can't change, but can be overridden?

### DISCUSSION THREAD

**Discussion by beauxq** [python/typing#1486] [Q&A] [answered]
`typing.Final` specifies that a value can't change, but it also specifies that the value can't be overridden in a subclass.
```
class PropertyBase:
    pass


class PropertyChild(PropertyBase):
    pass


class Base:
    x: ClassVar[PropertyBase] = PropertyBase()


class Child(Base):
    x: ClassVar[PropertyChild] = PropertyChild()  # error: Variable is mutable so its type is invariant...
```

It's good that pyright is reporting this error, because otherwise, when treating `Child` as `Base`, I could assign to its `x` a `PropertyBase` that is not `PropertyChild`, and invalidate the typing on `Child`.

But the problem is that I don't know how to specify it as immutable.
As long as I never change the `x` on `Base` and never change the `x` on `Child`, this should be valid and type safe.
But I don't know how to tell tell the type checker that `x` doesn't change, because `Final` doesn't only prevent it from changing, it also prevents it from being overridden.

```
class PropertyBase:
    pass


class PropertyChild(PropertyBase):
    pass


class Base:
    x: Final[PropertyBase] = PropertyBase()


class Child(Base):
    x: Final[PropertyChild] = PropertyChild()  # error: cannot be redeclared because parent class "Base" declares it as Final
```

How can I specify that the variable doesn't change, but it can be overridden in the subclass so that the tooling knows the more specific type?

**Comment by erictraut** [comment, marked as answer]
As you point out, `Final` implies both "immutable" and "can't be overridden". In your use case, you want to indicate that the field is read-only but still overridable by subclasses. There has been some discussion about adding [`ReadOnly`](https://discuss.python.org/t/pep-705-typedmapping/24827/17?u=erictraut) to the type system, but it hasn't gone very far.

This wasn't an issue prior to pyright enforcing [invariance for mutable fields](https://github.com/microsoft/pyright/issues/5933). Mypy users will also encounter the problem when [this feature](https://github.com/python/mypy/issues/3208) is implemented in mypy.

You may need to resort to a `# type: ignore` to work around this current shortcoming in the type system.
